go-controller/pkg/node/controllers/upgrade/node_upgrade.go

```diff
@@ -38,21 +40,31 @@ func NewController(client kubernetes.Interface, wf factory.NodeWatchFactory) *up
 
 // WaitForTopologyVerions polls continuously until the running master has reported a topology of
 // at least the minimum requested.
-func (uc *upgradeController) WaitForTopologyVersion(ctx context.Context, minVersion int, timeout time.Duration) error {
-	return wait.PollWithContext(ctx, 10*time.Second, timeout, func(ctx context.Context) (bool, error) {
-		ver, err := uc.GetTopologyVersion(ctx)
-		if err == nil {
-			if ver >= minVersion {
-				klog.Infof("Cluster topology version is now %d", ver)
-				return true, nil
-			}
+func (uc *upgradeController) WaitForTopologyVersion(ctx context.Context, stopChan <-chan struct{}, minVersion int, timeout time.Duration) error {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	timeoutContext, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	for {
+		select {
+		case <-ticker.C:
+			ver, err := uc.GetTopologyVersion(ctx)
+			if err == nil {
+				if ver >= minVersion {
+					klog.Infof("Cluster topology version is now %d", ver)
+					return nil
+				}
 
-			klog.Infof("Cluster topology version %d < %d", ver, minVersion)
-			return false, nil
+				klog.Infof("Cluster topology version %d < %d", ver, minVersion)
+			} else {
+				klog.Errorf("Failed to retrieve topology version: %v", err)
+			}
+		case <-stopChan:
+			return ErrStopped
+		case <-timeoutContext.Done():
+			return fmt.Errorf("failed to get topology version: %v", timeoutContext.Err())
```

Comment:
So now we have all these logs duplicated per network without really indicating which network is this for. 

Perhaps we could avoid klog here and let the caller do it. On timeout error, we should include the last topology version we were able to get in the error.

Reply:
but in the case of klog, it does not return to the caller yet.